File: shared/protocol.js
/*
  Shared Protocol Module
  ----------------------
  Defines all WebSocket message types shared between server and client.
  Messages are JSON-encoded with a fixed { type, payload } envelope.
*/

const MessageTypes = {
  // ─── Client → Server ──────────────────────────────────────────────────────

  // Navigate to a URL
  NAVIGATE: 'navigate',
  // { url: string, tabId?: string }

  // Mouse click (x, y are viewport pixel coordinates)
  CLICK: 'click',
  // { x: number, y: number, button: 'left'|'right'|'middle', tabId?: string }

  // Mouse down (start of drag)
  MOUSE_DOWN: 'mouseDown',
  // { x: number, y: number, button: 'left'|'right'|'middle', tabId?: string }

  // Mouse move (during drag or hover)
  MOUSE_MOVE: 'mouseMove',
  // { x: number, y: number, tabId?: string }

  // Mouse up (end of drag)
  MOUSE_UP: 'mouseUp',
  // { x: number, y: number, button: 'left'|'right'|'middle', tabId?: string }

  // Scroll by delta pixels
  SCROLL: 'scroll',
  // { deltaX: number, deltaY: number, tabId?: string }

  // Type text into the currently focused element
  TYPE: 'type',
  // { text: string, tabId?: string }

  // Press a single key
  KEY_PRESS: 'keyPress',
  // { key: string, modifiers?: { alt?: boolean, ctrl?: boolean, shift?: boolean, meta?: boolean }, tabId?: string }

  // Execute arbitrary JavaScript in the page
  EVALUATE: 'evaluate',
  // { code: string, tabId?: string }

  // Resize the browser viewport
  RESIZE: 'resize',
  // { width: number, height: number }

  // Take a fresh screenshot immediately
  REQUEST_SCREENSHOT: 'requestScreenshot',

  // Set proxy for current session
  SET_PROXY: 'setProxy',
  // { server: string, username?: string, password?: string }

  // Go back in history
  GO_BACK: 'goBack',

  // Go forward in history
  GO_FORWARD: 'goForward',

  // Create a new tab
  CREATE_TAB: 'createTab',
  // { url?: string }

  // Switch to a specific tab
  SWITCH_TAB: 'switchTab',
  // { tabId: string }

  // Close a tab
  CLOSE_TAB: 'closeTab',
  // { tabId: string }

  // Find text in page
  FIND_IN_PAGE: 'findInPage',
  // { text: string, tabId?: string }

  // ─── Server → Client ──────────────────────────────────────────────────────

  // Screen frame data (base64-encoded JPEG or PNG)
  FRAME: 'frame',
  // { data: string (base64), encoding: 'jpeg'|'png', width: number, height: number, tabId: string }

  // Current page URL changed
  URL_CHANGED: 'urlChanged',
  // { url: string, title: string, tabId: string }

  // Console messages from the page
  CONSOLE: 'console',
  // { type: string, args: string[], tabId: string }

  // Navigation error
  ERROR: 'error',
  // { message: string, tabId?: string }

  // Session info (sent on connect)
  SESSION_INFO: 'sessionInfo',
  // { sessionId: string, viewport: { width, height }, tabs: TabInfo[] }

  // Keepalive pong
  PONG: 'pong',

  // Page loading state changed
  LOADING_STATE: 'loadingState',
  // { loading: boolean, tabId?: string }

  // Updated tab list
  TAB_LIST: 'tabList',
  // { tabs: TabInfo[] }

  // Find in page results
  FIND_RESULTS: 'findResults',
  // { text: string, found: boolean, count: number, tabId?: string }
};


function encode(type, payload = {}) {
  return JSON.stringify({ type, payload, _t: Date.now() });
}


function decode(raw) {
  try {
    const msg = JSON.parse(raw);
    if (!msg.type || !MessageTypes[msg.type.toUpperCase()]) {
      return { type: 'unknown', payload: msg };
    }
    return msg;
  } catch {
    return { type: 'error', payload: { message: 'Invalid message format' } };
  }
}


module.exports = { MessageTypes, encode, decode };

File: client-rs/src/protocol.rs
/*
  Protocol — Shared Message Types (Rust)
  =======================================
  Mirrors the shared/protocol.js definitions. All WebSocket messages
  are JSON-encoded with { type, payload, _t: timestamp } envelope.

  This module provides type-safe structs for every message type
  exchanged between the Rust client and the Node.js server.
*/

use serde::{Deserialize, Serialize};

// ─── Envelope ───────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    #[serde(rename = "type")]
    pub msg_type: String,
    pub payload: serde_json::Value,
    #[serde(rename = "_t")]
    pub timestamp: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageEnvelope {
    pub msg_type: String,
    pub payload: serde_json::Value,
}

impl Message {
    pub fn new(msg_type: &str, payload: serde_json::Value) -> Self {
        Self {
            msg_type: msg_type.to_string(),
            payload,
            timestamp: Some(chrono_now()),
        }
    }

    pub fn to_json(&self) -> String {
        serde_json::to_string(self).unwrap_or_default()
    }

    pub fn from_json(json: &str) -> Option<Self> {
        serde_json::from_str(json).ok()
    }
}

fn chrono_now() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}

// ─── Message Type Constants ─────────────────────────────────────────────────

pub mod types {
    // Client → Server
    pub const NAVIGATE: &str = "navigate";
    pub const CLICK: &str = "click";
    pub const MOUSE_DOWN: &str = "mouseDown";
    pub const MOUSE_MOVE: &str = "mouseMove";
    pub const MOUSE_UP: &str = "mouseUp";
    pub const SCROLL: &str = "scroll";
    pub const TYPE: &str = "type";
    pub const KEY_PRESS: &str = "keyPress";
    pub const EVALUATE: &str = "evaluate";
    pub const RESIZE: &str = "resize";
    pub const REQUEST_SCREENSHOT: &str = "requestScreenshot";
    pub const SET_PROXY: &str = "setProxy";
    pub const GO_BACK: &str = "goBack";
    pub const GO_FORWARD: &str = "goForward";
    pub const CREATE_TAB: &str = "createTab";
    pub const SWITCH_TAB: &str = "switchTab";
    pub const CLOSE_TAB: &str = "closeTab";
    pub const FIND_IN_PAGE: &str = "findInPage";

    // Server → Client
    pub const FRAME: &str = "frame";
    pub const URL_CHANGED: &str = "urlChanged";
    pub const CONSOLE: &str = "console";
    pub const ERROR: &str = "error";
    pub const SESSION_INFO: &str = "sessionInfo";
    pub const PONG: &str = "pong";
    pub const EVALUATE_RESULT: &str = "evaluateResult";
    pub const LOADING_STATE: &str = "loadingState";
    pub const TAB_LIST: &str = "tabList";
    pub const FIND_RESULTS: &str = "findResults";
}

// ─── Payload Structs ────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NavigatePayload {
    pub url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClickPayload {
    pub x: u32,
    pub y: u32,
    #[serde(default = "default_button")]
    pub button: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

fn default_button() -> String {
    "left".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MouseDownPayload {
    pub x: u32,
    pub y: u32,
    #[serde(default = "default_button")]
    pub button: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MouseMovePayload {
    pub x: u32,
    pub y: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MouseUpPayload {
    pub x: u32,
    pub y: u32,
    #[serde(default = "default_button")]
    pub button: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScrollPayload {
    #[serde(default)]
    pub delta_x: i32,
    #[serde(default)]
    pub delta_y: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TypePayload {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeyPressPayload {
    pub key: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifiers: Option<KeyModifiers>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct KeyModifiers {
    #[serde(default, skip_serializing_if = "is_false")]
    pub alt: bool,
    #[serde(default, skip_serializing_if = "is_false")]
    pub ctrl: bool,
    #[serde(default, skip_serializing_if = "is_false")]
    pub shift: bool,
    #[serde(default, skip_serializing_if = "is_false")]
    pub meta: bool,
}

fn is_false(b: &bool) -> bool {
    !b
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResizePayload {
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SetProxyPayload {
    pub server: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub password: Option<String>,
}

// ─── Server → Client Payloads ──────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FramePayload {
    pub data: String,   // base64-encoded JPEG bytes
    pub encoding: String,
    pub width: u32,
    pub height: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UrlChangedPayload {
    pub url: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionInfoPayload {
    pub session_id: String,
    pub viewport: ViewportInfo,
    pub tabs: Vec<TabInfo>,
    #[serde(default)]
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViewportInfo {
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TabInfo {
    pub tab_id: String,
    pub url: String,
    pub title: String,
    pub active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SwitchTabPayload {
    pub tab_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FindInPagePayload {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoadingStatePayload {
    pub loading: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FindResultsPayload {
    pub text: String,
    pub found: bool,
    pub count: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ErrorPayload {
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tab_id: Option<String>,
}


/// Build a JSON payload for any serializable struct.
pub fn to_payload<T: Serialize>(val: &T) -> serde_json::Value {
    serde_json::to_value(val).unwrap_or(serde_json::Value::Null)
}
